### PR TITLE
refactor: Allocate IDs for tests

### PR DESCRIPTION
**Pull request by nayeemrmn**
Blocked on #14757.

Allows for better state-keeping of tests in Rust:
- Prints ` ... cancelled` for tests that were registered and cancelled due to uncaught errors:
  ```
  running 3 tests from ./temp.ts
  testWithUncaughtRejection ...
  Uncaught error from ./temp.ts FAILED
  testWithUncaughtRejection ... cancelled (0ms)
  anotherTest ... cancelled (0ms)
  anotherOne ... cancelled (0ms)
  ```
- Always know whether or not to reprint `someTest ...` when a result arrives (addresses https://github.com/denoland/deno/issues/14315#issuecomment-1138141249).
- Allows lots of future improvements to output e.g. more self-contained result lines for concurrent runs:
  ```
  (./temp.ts) t1 ... ok
  (./temp.ts) t2 ... t2-step1 ... FAILED
  (./temp.ts) t2 ... t2-step2 ... ok
  (./temp.ts) t2 ... FAILED
  ```

Adds `op_register_test()` to register tests, allocate IDs to them and potentially other info:
- Move filtering to rust, `{ id, filteredOut }` is returned by this op.
-  Tests are known about in Rust earlier, allowing for the `... cancelled` output above for tests that haven't started yet.

Cleanups in LSP test runner and more functional approach in `40_testing.js`. Bring bench runner error formatting up to date, it was still using `formatError()` in JS.

### CODE REVIEW

**Review by dsherret — APPROVED** [review]
LGTM. Probably @kitsonk should review the LSP changes before merge though.

**Review by bartlomieju — APPROVED** [review]
LGTM, nice work Nayeem. I verified that LSP changes work properly in vscode_deno.

File: cli/lsp/testing/collectors.rs
Comment (kitsonk) [inline cli/lsp/testing/collectors.rs:40]: Again, an arbitrary change.

File: cli/lsp/testing/definitions.rs
Comment (kitsonk) [inline cli/lsp/testing/definitions.rs:141]: some doc as to what this is any why

File: cli/lsp/testing/execution.rs
Comment (kitsonk) [inline cli/lsp/testing/execution.rs:64]: arbitrary renaming, why?  Our convention is that options are named "maybe_".

File: cli/lsp/testing/execution.rs
Comment (kitsonk) [inline cli/lsp/testing/execution.rs:125]: Arbitrary renaming, why?

File: cli/lsp/testing/execution.rs
Comment (kitsonk) [inline cli/lsp/testing/execution.rs:127]: Why is exclude now not optional?

File: cli/lsp/testing/execution.rs
Comment (kitsonk) [inline cli/lsp/testing/execution.rs:130]: arbitrary renaming, why?

File: cli/lsp/testing/execution.rs
Comment (kitsonk) [inline cli/lsp/testing/execution.rs:994]: Why remove a test assertion?  Specifically testing the serialization of JSON, where else is this tested?

File: cli/lsp/testing/lsp_custom.rs
Comment (kitsonk) [inline cli/lsp/testing/lsp_custom.rs:98]: Changing things like this that come from the client changes the API with the client.  This message is specifically sent from the client, and you are now making an optional property mandatory. 😢   Why make a breaking change with the API?

File: cli/lsp/testing/lsp_custom.rs
Comment (kitsonk) [inline cli/lsp/testing/lsp_custom.rs:26]: This changes the deserialization of this structure.  I can't remember off hand if this is ever sent from the client, but why arbitrarily break the contract with the client?

**Review by kitsonk — COMMENTED** [review]
I'm concerned some of the changes to the lsp aren't "cleanup" but arbitrary changes.  Specifically for the messages that go to and from the client that are changed should have their types changed in `https://github.com/denoland/vscode_deno/blob/main/client/src/lsp_extensions.ts` to ensure there isn't a knock on effect.  Especially when we remove `Option` it becomes a mandatory field on deserializing.

File: cli/lsp/testing/collectors.rs
Comment (nayeemrmn) [reply cli/lsp/testing/collectors.rs:18]: This is a redundant `Option` unless I missed something, if there are no steps that is represented as an empty vec

File: cli/lsp/testing/execution.rs
Comment (nayeemrmn) [reply cli/lsp/testing/execution.rs:127]: This is another redundant `Option`, blocklists are inherently non-optional because that's the same thing as being empty

File: cli/lsp/testing/execution.rs
Comment (nayeemrmn) [reply cli/lsp/testing/execution.rs:64]: That's not true, I saw https://github.com/denoland/deno/blob/7d9e8baebb6c2534346e0823e2cc6408ebb8b1eb/cli/lsp/testing/lsp_custom.rs#L90-L99 and plenty of other places. It was attractive to switch here because `exclude` was made non-optional.

In my experience, prefixes like `try_foo()` and `maybe_foo()` are used for _methods_ to distinguish from their counterparts which panic on failure/absence. It's usually not needed for fields and variables thanks to editors, it's sort of Hungarian

File: cli/lsp/testing/execution.rs
Comment (nayeemrmn) [reply cli/lsp/testing/execution.rs:125]: There is another `TestFilter` added in `cli/tools/test.rs`, it's used in this file

File: cli/lsp/testing/execution.rs
Comment (nayeemrmn) [reply cli/lsp/testing/execution.rs:994]: `.as_test_options()` doesn't exist anymore, filters aren't passed to JS since it's moved to Rust

File: cli/lsp/testing/lsp_custom.rs
Comment (nayeemrmn) [reply cli/lsp/testing/lsp_custom.rs:98]: See https://serde.rs/field-attrs.html#default

File: cli/lsp/testing/definitions.rs
Comment (nayeemrmn) [reply cli/lsp/testing/definitions.rs:141]: Sure, though it's already clearer than before... we had `TestDefinitions::injected` already, I extracted a method to populate it. I suppose the return-if-insertion-succeeded thing could be doc'd

File: cli/tests/testdata/test/uncaught_errors.out
Comment (dsherret) [inline cli/tests/testdata/test/uncaught_errors.out:56]: @nayeemrmn seems like this is flaky now: https://github.com/denoland/deno/runs/7362087847?check_suite_focus=true

**Pull request merged**
